### ISSUE TITLE
iOS only SocketException No route to host

### ISSUE THREAD

**Issue by CsabaStupak** [dotnet/maui#14224]
### Description

I have a simple Maui application (.NET7 - created from template) which communicates with a local network server. I'm testing it using Android and iPhone. The code is simple:

```
TcpClient client = new TcpClient();
await client.ConnectAsync(IPAddress.Parse("192.168.3.10"), 12345);
```

or another approach:

```
Ping ping = new Ping();
await ping.SendPingAsync(IPAddress.Parse("192.168.3.10"));
```

In both case the result is _SocketException (65): No route to host_. This happens on iPhone only. On Android it works fine.

My Android phone IP is 192.168.3.45 and iPhone IP is 192.168.3.170. Both are on the same network as the server. First time I try to connect even the "Grant access to local network" appears on iPhone, what I allowed. I can see in the iPhone settings that my application has access to _local network = on_.

In the project I even added the following to the Info.plist:

```
<key>NSAppTransportSecurity</key>
<dict>
    <key>NSAllowsLocalNetworking</key>
    <true/>
    <key>NSAllowsArbitraryLoads</key>
    <true/>
</dict>
```

On my **iPhone 13 mini**, **iOS 16.3.1** I can see the server using _Fing - Network Scanner_ and its _Device Discovery_. I'm able to Ping it. So my phones are certainly on the same network. Since I can connect to the server via Android my server settings should be OK (firewall etc.). The given code works fine if I try to ping 8.8.8.8 address. So this is apparently a local network access issue - somehow iPhone ignores that setting.

Any idea what could be the problem? Thanks :-)

### Steps to Reproduce

1. Create a New .NET MAUI app or clone the following repository https://github.com/CsabaStupak/MauiLocNetIssue.git
2. Add the above code in try/catch block to the `OnCounterClicked` method.
3. Modify the Info.plist file.
4. Connect to your local iPhone device and click the button and in the debugger see the exception thrown.

### Link to public reproduction project repository

https://github.com/CsabaStupak/MauiLocNetIssue.git

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iOS 16.3.1

### Did you find any workaround?

No workaround found and since our application is a relay between Bluetooth device and local network server so can't move forward :-(

### Relevant log output

```shell
[0:] Ping to 192.168.3.10 from 192.168.3.170 failed: System.Net.NetworkInformation.PingException: An exception occurred during a Ping request.
 ---> System.Net.Sockets.SocketException (65): No route to host
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.CreateException(SocketError error, Boolean forAsyncThrow)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.SendToAsync(Socket socket, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.SendToAsync(ReadOnlyMemory`1 buffer, SocketFlags socketFlags, EndPoint remoteEP, CancellationToken cancellationToken)
   at System.Net.NetworkInformation.Ping.SendIcmpEchoRequestOverRawSocketAsync(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[<SendIcmpEchoRequestOverRawSocketAsync>d__70](<SendIcmpEchoRequestOverRawSocketAsync>d__70& stateMachine)
   at System.Net.NetworkInformation.Ping.SendIcmpEchoRequestOverRawSocketAsync(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options)
   at System.Net.NetworkInformation.Ping.SendPingAsyncCore(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[<SendPingAsyncCore>d__84](<SendPingAsyncCore>d__84& stateMachine)
   at System.Net.NetworkInformation.Ping.SendPingAsyncCore(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options)
   at System.Net.NetworkInformation.Ping.SendPingAsyncInternal(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[<SendPingAsyncInternal>d__51](<SendPingAsyncInternal>d__51& stateMachine)
   at System.Net.NetworkInformation.Ping.SendPingAsyncInternal(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options)
   at System.Net.NetworkInformation.Ping.SendPingAsync(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options)
   at System.Net.NetworkInformation.Ping.SendPingAsync(IPAddress address)
   at MauiApp7.MainPage.PingAsync(String ipAddr) in D:\Data\DF19\Garmin\Test\TestNet7\MauiApp7\MainPage.xaml.cs:line 44
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[<PingAsync>d__3](<PingAsync>d__3& stateMachine)
   at MauiApp7.MainPage.PingAsync(String ipAddr)
   at MauiApp7.MainPage.OnCounterClicked(Object sender, EventArgs e) in D:\Data\DF19\Garmin\Test\TestNet7\MauiApp7\MainPage.xaml.cs:line 36
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[MauiApp7.MainPage.<OnCounterClicked>d__2, MauiApp7, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[MauiApp7.MainPage.<OnCounterClicked>d__2, MauiApp7, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[MauiApp7.MainPage.<OnCounterClicked>d__2, MauiApp7, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
   at System.Threading.Tasks.AwaitTaskContinuation.<>c.<.cctor>b__17_0(Object state)
   at System.Threading.Tasks.AwaitTaskContinuation.RunCallback(ContextCallback callback, Object state, Task& currentTask)
   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.Run(Task task, Boolean canInlineContinuationTask)
   at System.Threading.Tasks.Task.RunContinuations(Object continuationObject)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].TrySetResult(VoidTaskResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].SetExistingTaskResult(Task`1 task, VoidTaskResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()
   at MauiApp7.MainPage.PingAsync(String ipAddr) in D:\Data\DF19\Garmin\Test\TestNet7\MauiApp7\MainPage.xaml.cs:line 53
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Runtime.CompilerServices.IAsyncStateMachine, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Runtime.CompilerServices.IAsyncStateMachine, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext(Thread threadPoolThread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.Runtime.CompilerServices.IAsyncStateMachine, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].MoveNext()
   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.<>c.<.cctor>b__8_0(Object state)
   at Foundation.NSAsyncSynchronizationContextDispatcher.Apply()
   at UIKit.UIApplication.xamarin_UIApplicationMain(Int32 , String[] , IntPtr , IntPtr , IntPtr& )
   at UIKit.UIApplication.UIApplicationMain(Int32 argc, String[] argv, IntPtr principalClassName, IntPtr delegateClassName)
   at UIKit.UIApplication.Main(String[] args, Type principalClass, Type delegateClass)
   at MauiApp7.Program.Main(String[] args) in D:\Data\DF19\Garmin\Test\TestNet7\MauiApp7\Platforms\iOS\Program.cs:line 13
   at System.Reflection.RuntimeMethodInfo.InternalInvoke(RuntimeMethodInfo , Object , Span`1& , Exception& )
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Xamarin.iOS.HotRestart.Application.Run(String[] args, Action`1 loadFrameworkAction, Action killApplicationAction)
   at Xamarin.PreBuilt.iOS.Applications.Main(String[] args) in /Users/builder/azdo/_work/1/s/xamarin-macios/msbuild/Xamarin.HotRestart.PreBuilt/Xamarin.PreBuilt.iOS/Main.cs:line 12
--- End of stack trace from previous location ---
   at System.Net.NetworkInformation.Ping.SendIcmpEchoRequestOverRawSocketAsync(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options)
   at System.Net.NetworkInformation.Ping.SendPingAsyncCore(IPAddress address, Byte[] buffer, Int32 timeout, PingOptions options)
   at System.Net.NetworkInformation.Ping.SendPingAsyncInternal(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options)
   --- End of inner exception stack trace ---
   at System.Net.NetworkInformation.Ping.SendPingAsyncInternal(IPAddress address, Int32 timeout, Byte[] buffer, PingOptions options)
   at MauiApp7.MainPage.PingAsync(String ipAddr) in D:\Data\DF19\Garmin\Test\TestNet7\MauiApp7\MainPage.xaml.cs:line 44
```

**Comment by jfversluis** [comment]
If you create a iOS app without .NET MAUI (do `dotnet new ios`), does it work then? I kind of assume it doesn't. I don't think this is specific to .NET MAUI.

**Comment by CsabaStupak** [comment]
@jfversluis I'm a Windows developer I can't try the IOS only project. I own only an iPhone. The project seems to run only on Mac so don't know how to verify if this is the same case.

**Comment by jfversluis** [comment]
Alright I tried it for you. Both with an iOS project without .NET MAUI and with a .NET MAUI project, running on an iOS Simulator _and_ a physical device I can't reproduce this issue. Then tried your specific reproduction and also that one runs fine.

![image](https://user-images.githubusercontent.com/939291/228267190-930adb1d-7eee-4e96-96a0-b7972bb4c409.png)

I only modified the IP address to be one on my local network. The exception you see is to be expected since I don't have anything running on TCP port 4356.

So a couple of options come to mind:
* Something isn't right from inside your network
* You're hitting a limitation when using Hot Reload (the name we have for developing for iOS devices directly from Windows)

I'm leaning towards the first one since it does throw an exception that has a relevant message: no route to host. This indicates to me that the ping/connection operation is actually attempted, but fails. In both cases I don't think there is much we can do here and it will be pretty hard to diagnose on our end.

Closing this for now as this doesn't seem like a bug with us from what I can tell. Good luck!
